### PR TITLE
chore(search): no need to normalize params for search during init

### LINKED ISSUE #11483: master: search all not finding groups, but searching specific does
Maybe not related to groups, but found out there

/cc @hypeJunction

### PR DESCRIPTION
**Pull request by jdalsem**
fixes #11483

### CODE REVIEW

File: engine/classes/Elgg/Search/SearchService.php [hunk 1]
```diff
@@ -234,7 +234,7 @@ public function normalizeSearchFields(array $options = []) {
 		} else {
 			// only allow known fields
 			foreach ($fields as $property_type => $property_type_fields) {
-				if (empty($options['fields'][$property_type])) {
+				if (empty($property_type_fields)) {
```
Comment (hypeJunction) [inline engine/classes/Elgg/Search/SearchService.php]: I don't think that's what's intended here. Let's see, but tests will likely fail

Comment (jdalsem) [reply engine/classes/Elgg/Search/SearchService.php]: i believe it is... you will want to reset the allowed fields if non are allowed/known

Comment (hypeJunction) [reply engine/classes/Elgg/Search/SearchService.php]: That's right, but it should happen before this loop. This loops normalizes the known fields.

Comment (jdalsem) [reply engine/classes/Elgg/Search/SearchService.php]: tests are failing as expected, but maybe the test assume the wrong thing...

this foreach should do the following if i am correct:
if provided `$options['fields'][$type]` contains fields that are not present in the allowed `$fields` they should be removed... thus if `$fields[$type]`  is empty `$options['fields'][$type]` should also be empty...

Comment (hypeJunction) [reply engine/classes/Elgg/Search/SearchService.php]: There is difference between empty and not set. If the fields are not set, we default to known fields, if they are set as false or [], then we don't use any.

**Comment by jdalsem** [comment]
hmm the search plugin test fail because the tests assume too much

File: engine/classes/Elgg/Search/SearchService.php [hunk 2]
```diff
@@ -234,12 +234,12 @@ public function normalizeSearchFields(array $options = []) {
 		} else {
 			// only allow known fields
 			foreach ($fields as $property_type => $property_type_fields) {
-				if (empty($options['fields'][$property_type])) {
+				if (empty($property_type_fields) || empty($options['fields'][$property_type])) {
```
Comment (jdalsem) [inline engine/classes/Elgg/Search/SearchService.php]: both situations result in the quick return

File: engine/classes/Elgg/Search/SearchService.php [hunk 3]
```diff
@@ -234,12 +234,12 @@ public function normalizeSearchFields(array $options = []) {
...
 					$options['fields'][$property_type] = [];
 					continue;
 				}
 
-				$allowed = array_intersect($property_type_fields, (array) $options['fields'][$property_type]);
+				$allowed = array_intersect((array) $options['fields'][$property_type], $property_type_fields);
```
Comment (jdalsem) [inline engine/classes/Elgg/Search/SearchService.php]: result is the same, but this is the correct logical order of attributes

File: mod/search/classes/Elgg/Search/Highlighter.php [hunk 4]
```diff
@@ -46,7 +46,7 @@ public function highlightWords($text) {
 			'search-highlight-color' => rand(10000, 99999)
 		];
 
-		$parts = elgg_extract('query_parts', $this->params);
+		$parts = (array) elgg_extract('query_parts', $this->params, []);
```
Comment (jdalsem) [inline mod/search/classes/Elgg/Search/Highlighter.php]: do not assume query parts exist

File: mod/search/tests/phpunit/unit/Elgg/Search/SearchPluginTest.php [hunk 5]
```diff
@@ -35,17 +35,15 @@ public function testParameterInitialization() {
 
 		$this->assertEquals([
 			'query' => 'literal needle',
-			'query_parts' => ['literal needle'],
```
Comment (jdalsem) [inline mod/search/tests/phpunit/unit/Elgg/Search/SearchPluginTest.php]: can not test this as it is not part of the plugin... assuming the provider will create them is incorrect

File: mod/search/tests/phpunit/unit/Elgg/Search/SearchPluginTest.php [hunk 6]
```diff
@@ -35,17 +35,15 @@ public function testParameterInitialization() {
...
 			'limit' => 10,
 			'sort' => 'time_created',
 			'order' => 'desc',
 			'search_type' => 'custom',
 			'fields' => [
-				'metadata' => [],
-				'annotations' => [],
-				'private_settings' => [],
-				'attributes' => [],
+				'metadata' => [
```
Comment (jdalsem) [inline mod/search/tests/phpunit/unit/Elgg/Search/SearchPluginTest.php]: only test what the search plugins search helper class determines

File: mod/search/tests/phpunit/unit/Elgg/Search/SearchPluginTest.php (hunk 5 above)
Comment (hypeJunction) [reply mod/search/tests/phpunit/unit/Elgg/Search/SearchPluginTest.php]: It is in the search plugin parameter normalization

Comment (jdalsem) [reply mod/search/tests/phpunit/unit/Elgg/Search/SearchPluginTest.php]: query parts is generated by the service, not by the search plugin

Comment (hypeJunction) [reply mod/search/tests/phpunit/unit/Elgg/Search/SearchPluginTest.php]: search plugin checks if the search string is quoted and sets to tokenization to false

Comment (jdalsem) [reply mod/search/tests/phpunit/unit/Elgg/Search/SearchPluginTest.php]: based on query, not on query_parts... i only removed the query_parts... search plugin does not know what the effect is of tokenization

**Comment by hypeJunction** [comment]
This can be closed

**Pull request closed without merging**